### PR TITLE
fix for issue about attempted deserialization of multipart form data …

### PR DESCRIPTION
**Pull request by emil-perhinschi**
… when a serializer is set

a quick solution for issue 1677

### CODE REVIEW

**Comment by cromedome** [comment]
Hi Emil,

Thanks for reporting the bug and for your PR!

I see what you did, and I think it's a fair approach to the problem, but in running the Dancer2 test suite against your proposed fix, it produces a *lot* of uninitialized warning messages that weren't present before. Could you please run your fix against our test suite and resolve the warnings? I'll take another look once the warnings are handled.

Thanks so much!

**Comment by emil-perhinschi** [comment]
> in running the Dancer2 test suite against your proposed fix, it produces a _lot_ of uninitialized warning messages that weren't present before.

Hello Chromedome, 

I did run `prove -v` inside the lastest Dancer2 `master` branch and there were no uninitialized warnings. 
sftp://192.168.100.51/home/emilper/personal/Dancer2/prove_verbose.log

I have attached the log.

**Comment by emil-perhinschi** [comment]
Hello Jason, 

you mean the uninitialized errors in https://ci.appveyor.com/project/xsawyerx/dancer2/builds/46335065 ? 

Can't find the file 00-report-prereqs.t in master ... where can I find/get those tests ? 

thank you,

Emil

**Comment by emil-perhinschi** [comment]
found them, ran, made the fix, ran `dzil test` again, pushed 

Just in case, there are a few modules that are missing in `cpanfile` and not installed by Dist::Zilla: 
Test::CPAN::Meta
Dancer2::Session::Cookie
Test::NoTabs
which appear to be needed in tests

**Comment by emil-perhinschi** [comment]
Is there another problem with my fix ? AppVeyor build succeeded .

**Review by cromedome — APPROVED** [review]
I would like to see a test to show that the bug is fixed. Nonetheless, this gets a :+1:  from me.

**Comment by emil-perhinschi** [comment]
updated the pull request with a test

**Comment by cromedome** [comment]
@emil-perhinschi Awesome. I ran your test without your fixes applied to verify the bug, and again with them to confirm everything works properly.  This gets a big :+1:  from me. 

@xsawyerx? @yanick? Penny for your thoughts!

File: lib/Dancer2/Core/Request.pm [hunk 1]
```diff
@@ -181,6 +181,18 @@ sub data { $_[0]->{'data'} ||= $_[0]->deserialize() }
 sub deserialize {
     my $self = shift;
 
+    # if a serializer is used (JSON or Mutable) the deserialization of 
+    #   the whole 'multipart' section of the request is attempted
+    #   starting with the boundary
+    if (
+        $self->content_type 
+        && $self->content_type =~ /^multipart\/form-data/i 
+        && $self->serializer
```
Comment (yanick) [inline lib/Dancer2/Core/Request.pm:190]: I think this last condition is not necessary. For one it's taken care of at line 196 right below, and for another I think the logical condition here is "no matter if we have a serializer or not, we don't no nuthing if the content type is `multipart/form-data`. Nuthing, I tell you.

**Review by yanick — CHANGES_REQUESTED** [review]
I have this one comment I left, but beside that, looks good!

File: lib/Dancer2/Core/Request.pm [hunk 2]
```diff
@@ -181,6 +181,18 @@ sub data { $_[0]->{'data'} ||= $_[0]->deserialize() }
 sub deserialize {
     my $self = shift;
 
+    # if a serializer is used (JSON or Mutable) the deserialization of
```
Comment (yanick) [inline lib/Dancer2/Core/Request.pm:184]: I think the comment describe the problem the code fixes. But without context I am prone to think it describes what the block does. 

Maybe `# don't attempt to deserialize if the form is 'multipart/form-data'`

**Comment by emil-perhinschi** [comment]
updated as requested

**Review by yanick — APPROVED** [review]

**Comment by cromedome** [comment]
This was merged but never closed. Thanks again! :)

**Comment by emil-perhinschi** [comment]
thank you

will it make into a CPAN release ? else we can continue using github and manual installs, but that's not the optimum solution :)

**Comment by cromedome** [comment]
It's queued up for the next release, tentatively next week.

**Pull request closed without merging**